server/datastore/mysql/microsoft_mdm.go

```diff
@@ -90,3 +90,187 @@ func (ds *Datastore) MDMWindowsDeleteEnrolledDevice(ctx context.Context, mdmDevi
 
 	return ctxerr.Wrap(ctx, notFound("MDMWindowsEnrolledDevice"))
 }
+
+// MDMWindowsDeletePendingCommands deletes pending commands for a given device id.
+func (ds *Datastore) MDMWindowsDeletePendingCommands(ctx context.Context, deviceID string) error {
+	stmt := "DELETE FROM windows_mdm_pending_commands WHERE device_id = ?"
+
+	_, err := ds.writer(ctx).ExecContext(ctx, stmt, deviceID)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "delete MDMWindowsDeletePendingCommands")
+	}
+
+	return nil
+}
+
+// MDMWindowsInsertPendingCommand inserts a new WindowMDMPendingCommand in the database
+func (ds *Datastore) MDMWindowsInsertPendingCommand(ctx context.Context, cmd *fleet.MDMWindowsPendingCommand) error {
+	stmt := `
+		INSERT INTO windows_mdm_pending_commands (
+		command_uuid,
+		device_id,
+		cmd_verb,
+		setting_uri,
+		setting_value,
+		data_type,
+		system_origin ) VALUES (?, ?, ?, ?, ?, ?, ?)
+	`
+	_, err := ds.writer(ctx).ExecContext(
+		ctx,
+		stmt,
+		cmd.CommandUUID,
+		cmd.DeviceID,
+		cmd.CmdVerb,
+		cmd.SettingURI,
+		cmd.SettingValue,
+		cmd.DataType,
+		cmd.SystemOrigin)
+	if err != nil {
+		if isDuplicate(err) {
+			return ctxerr.Wrap(ctx, alreadyExists("MDMWindowsPendingCommand", cmd.CommandUUID))
+		}
+		return ctxerr.Wrap(ctx, err, "inserting MDMWindowsPendingCommand")
+	}
+
+	return nil
+}
+
+// MDMWindowsGetPendingCommands retrieves all commands for a given device ID from the windows_mdm_pending_commands table
+func (ds *Datastore) MDMWindowsGetPendingCommands(ctx context.Context, deviceID string) ([]*fleet.MDMWindowsPendingCommand, error) {
+	var commands []*fleet.MDMWindowsPendingCommand
+
+	query := `
+        SELECT
+            command_uuid,
+            device_id,
+            cmd_verb,
+            setting_uri,
+            setting_value,
+            data_type,
+            system_origin,
+            created_at,
+            updated_at
+        FROM
+            windows_mdm_pending_commands
+        WHERE
+            device_id = ?
+    `
+
+	// Retrieve commands first
+	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &commands, query, deviceID); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "get pending Windows MDM commands by device id")
+	}
+
+	// Then check which commands are not currently tracked in the windows_mdm_commands table
+	trackedCmds, err := ds.MDMWindowsListCommands(ctx, deviceID)
```

Comment:
THat would be best done in the query itself, you can get those from `windows_mdm_pending_commands` that don't exist in `windows_mdm_commands` by doing something like this (untested, but that's the general idea):

```
SELECT 
  ...
FROM 
   windows_mdm_pending_commands wmpc
WHERE
  wmpc.device_id = ? AND
  NOT EXISTS (SELECT 1 FROM windows_mdm_commands wmc WHERE wmpc.device_id = wmc.device_id AND wmpc.command_uuid = wmc.command_uuid)
```

Reply:
Thanks, I used the query to retrieve the filtered cmds instead of programmatic logic